### PR TITLE
LEDC - AnalogWrite new API + ledcAttachPin duty fix

### LINKED ISSUE #7105: PWM: Further analogWrite broken after ledcDetachPin
### Board

ESP32 Dev Module

### Device Description

Random Devkit

### Hardware Configuration

Any GPIO than can do PWM

### Version

v2.0.4

### IDE Name

Arduino IDE

### Operating System

Whatever

### Flash frequency

Whatever

### PSRAM enabled

no

### Upload speed

Whatever

### Description

The following sequence of function calls 

analogWrite(pin, 128);
ledcDetachPin(pin);
analogWrite(pin,128);

does only give PWM once and not after the second call to analogWrite.

Problem: Internal conditional on pin_to_channel array prevents second ledcAttachPin.

Solution:
```
$ diff -u .arduino15/packages/esp32/hardware/esp32/2.0.4/cores/esp32/esp32-hal-ledc.c{.orig,}
--- .arduino15/packages/esp32/hardware/esp32/2.0.4/cores/esp32/esp32-hal-ledc.c.orig	2022-07-06 11:47:49.000000000 +0200
+++ .arduino15/packages/esp32/hardware/esp32/2.0.4/cores/esp32/esp32-hal-ledc.c	2022-08-10 00:18:47.778755484 +0200
@@ -222,6 +222,8 @@
       pin_to_channel[pin] = cnt_channel--;
       ledcAttachPin(pin, cnt_channel);
       ledcSetup(cnt_channel, 1000, 8);
+    } else {
+      ledcAttachPin(pin, cnt_channel);
     }
     ledcWrite(pin_to_channel[pin] - 1, value);
   }
```

Apply inline patch and it works.

Regards,
Harald.

### Sketch

```cpp
See above, you can reproduce
```

### Debug Message

```plain
See above, you can reproduce
```

### Other Steps to Reproduce

_No response_

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### LINKED ISSUE #7306: LEDC pull request introduced regressions https://github.com/espressif/arduino-esp32/pull/6045
### Board

ESP32 Dev Module

### Device Description

Dev board and LED

### Hardware Configuration

GPIO connected to LED

### Version

v2.0.4

### IDE Name

Arduino IDE

### Operating System

Linux

### Flash frequency

Whatever

### PSRAM enabled

no

### Upload speed

115200

### Description

The pull request https://github.com/espressif/arduino-esp32/pull/6045 introduced two regressions:

1. It is not possible any more to do the following any more:

ledcSetup(...); ledcAttachPin(...); ledcWrite(...); // set up channel with freqency an duty cycle to blink LED
ledcDetachPin(...) // remove pin (LED OFF)
ledcAttachPin(...) // attach pin again and it continues with SAME frequency and duty as before

as ledcAttachPin nowadays resets the duty cycle of a prefectly working channel setup.

2. ledcSetup(channel, 1, 8) can nor set up frequency of 1Hz any more

Harald.

### Sketch

```cpp
I think you understand anyway, see above.
```

### Debug Message

```plain
00:49:16.724 -> E (247241) ledc: requested frequency and duty resolution can not be achieved, try reducing freq_hz or duty_resolution. div_param=8460032
00:49:16.758 -> [247245][E][esp32-hal-ledc.c:75] ledcSetup(): ledc setup failed!
```

### Other Steps to Reproduce

_No response_

### I have checked existing issues, online documentation and the Troubleshooting Guide

- [X] I confirm I have checked existing issues, online documentation and Troubleshooting guide.

### PR DESCRIPTION
**Pull request by P-R-O-C-H-Y**
## Description of Change
1. Function `ledcAttachPin()` now reads duty from channel before setup in order to keep its current duty.
2. Added new API functions for analogWrite, which sets frequency or resolution for all led channels used by analogWrite.

- `analogWriteFrequency(uint32_t freq)`
- `analogWriteResolution(uint8_t bits)`

## Tests scenarios
Tested on ESP32.

Sketch for testing the fix with analogWrite(connect led to pin 2 - should blink with 500ms delays):

```
uint8_t led = 2;

void setup() 
{            
  analogWrite(led, 128);
  delay(500);
}

void loop() 
{
  ledcDetachPin(led);
  delay(500);
  ledcAttachPin(led,analogGetChannel(led));
  analogWrite(led,128);
  delay(500);
}
```

## Related links
Closes #7105 
Closes #7306

### CODE REVIEW

**Comment by PilnyTomas** [comment]
Tested on ESP32 with following sketch and works ok - signal is not terminated
```
void setup() {
  uint8_t pin = 14;
  analogWrite(pin, 128);
  delay(50);
  ledcDetachPin(pin);
  delay(50);
  ledcAttachPin(pin,analogGetChannel(pin)); //added
  delay(50);
  analogWrite(pin,128);
}
void loop() {}

```

**Review by me-no-dev — APPROVED** [review]

File: cores/esp32/esp32-hal-ledc.c [hunk 1]
```diff
@@ -230,3 +233,23 @@ void analogWrite(uint8_t pin, int value) {
 int8_t analogGetChannel(uint8_t pin) {
     return pin_to_channel[pin] - 1;
 }
+
+void analogWriteFrequency(uint32_t freq) {
+    if (cnt_channel == LEDC_CHANNELS) {
+        return; //No channel used for analogWrite
+    }
+    for (int channel = LEDC_CHANNELS - 1; channel >= cnt_channel; channel--) {
```
[the remark this answers is not in the document]
Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-ledc.c:241]: IDF is checking the frequency input and validating if its valid to set or not. Same with resolution. IDF will return error that is checked in Arduino.

File: cores/esp32/esp32-hal-ledc.c [hunk 2]
```diff
@@ -230,3 +233,23 @@ void analogWrite(uint8_t pin, int value) {
...
+    }
+    analog_frequency = freq;
+}
+
+void analogWriteResolution(uint8_t bits) {
+    if (cnt_channel == LEDC_CHANNELS) {
+        return; //No channel used for analogWrite
+    }
+    for (int channel = LEDC_CHANNELS - 1; channel >= cnt_channel; channel--) {
+        ledcChangeFrequency(channel, analog_frequency, bits);
```
[the remark this answers is not in the document]
Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-ledc.c:252]: You will get an error that is not valid input.

[the remark this answers is not in the document]
Comment (SuGlider) [reply cores/esp32/esp32-hal-ledc.c:252]: I tried that and I got no message. 
Are you sure that changing resolution is done with `ledcChangeFrequency()` call?
It doesn't call `ledc_timer_config()`
I think it is changing the Duty only, not the resolution.

File: cores/esp32/esp32-hal-ledc.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (SuGlider) [reply cores/esp32/esp32-hal-ledc.c:241]: I tested it with wrong values and I got no error message.
Looking into https://github.com/espressif/esp-idf/blob/release/v4.4/components/driver/ledc.c#L558
I see duty validation, but it doesn't fail...

But if I test it with IDF I get the error...

File: cores/esp32/esp32-hal-ledc.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (SuGlider) [reply cores/esp32/esp32-hal-ledc.c:252]: I tested it with wrong value and I got no error message.
Looking into https://github.com/espressif/esp-idf/blob/release/v4.4/components/driver/ledc.c#L558
I see duty validation, but it doesn't fail...

But if I test it with IDF I get the error...

**Comment by SuGlider** [comment]
@P-R-O-C-H-Y  - Please try this sketch:

``` cpp
#include <driver/ledc.h>

void setup() {
  uint8_t chan = 0;
  uint8_t bit_num = 100;
  uint32_t freq = 1000000000;
  uint8_t group = (chan / 8), timer = ((chan / 2) % 4);

  ledc_timer_config_t ledc_timer = {
    .speed_mode       = (ledc_mode_t) group,
    .duty_resolution  = (ledc_timer_bit_t) bit_num,
    .timer_num        = (ledc_timer_t) timer,
    .freq_hz          = freq,
    .clk_cfg          = LEDC_AUTO_CLK //LEDC_DEFAULT_CLK
  };

  if (ledc_timer_config(&ledc_timer) != ESP_OK)
  {
    log_e("ledc setup failed!");
  } else {
    log_e("ledc setup ALL RIGHT!");
  }


  // no error happens in setup!
  analogWriteResolution(128);
  analogWriteFrequency(1000000000UL);
}

void loop() {
}
```


### This is the output I get... With IDF it issues an error, but with Arduino Layer... nothing!

``` 
ets Jun  8 2016 00:22:57

rst:0x1 (POWERON_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
flash read err, 1000
ets_main.c 371 
ets Jun  8 2016 00:22:57

rst:0x10 (RTCWDT_RTC_RESET),boot:0x13 (SPI_FAST_FLASH_BOOT)
configsip: 0, SPIWP:0xee
clk_drv:0x00,q_drv:0x00,d_drv:0x00,cs0_drv:0x00,hd_drv:0x00,wp_drv:0x00
mode:DIO, clock div:1
load:0x3fff0030,len:1344
load:0x40078000,len:13864
load:0x40080400,len:3608
entry 0x400805f0
[     3][D][esp32-hal-cpu.c:244] setCpuFrequencyMhz(): PLL: 480 / 2 = 240 Mhz, APB: 80000000 Hz
E (30) ledc: freq_hz=1000000000 duty_resolution=100
[    32][E][Fading.ino:39] setup(): ledc setup failed!

```

**Comment by SuGlider** [comment]
I think it has to do with the order of the initialization of the fields in C structs...

**Comment by SuGlider** [comment]
OK... the issue is because of 
`static int cnt_channel = LEDC_CHANNELS;` that never changes...

So if the sketch never calls `analogWrite()` before calling `analogWritResolution()`, no channel is allocated and it does nothing.

I think this may not be the expected behaviour. The example from Arduino Github says it should work.
https://github.com/arduino/reference-tr/blob/master/Language/Functions/Zero%2C%20Due%2C%20MKR%20Family/analogWriteResolution.adoc

**Comment by P-R-O-C-H-Y** [comment]
> OK... the issue is because of `static int cnt_channel = LEDC_CHANNELS;` that never changes...
> 
> Thus, if the sketch never calls `analogWrite()` before calling `analogWritResolution()`, no channel is allocated and it does nothing.
> 
> I think this isn't the expected behaviour. The example from Arduino Github says it should work. https://github.com/arduino/reference-tr/blob/master/Language/Functions/Zero%2C%20Due%2C%20MKR%20Family/analogWriteResolution.adoc
> 
> Please read the notes at the end of the document from Arduino.cc - it says what should happen when the resolution parameter is higher than what the device can handle. We shall do the same.
> 
> Another point is that the resolution and frequency shall be applied to all channels/pins. https://github.com/arduino/ArduinoCore-sam/blob/master/cores/arduino/wiring_analog.c There you can see that there is only one resolution to every Analog output pin.
> 
> I think we shall try to comply always with Arduino MainStream.

The problem is, that its our analogWrite is using LEDC. I think its not good to change Resolution or Frequency on all LEDC channels when not used by analogWrite. That can override some already setup led channels that are not used by analogWrite. Where I see a mistake is that return.
So we can change this:
```
void analogWriteResolution(uint8_t bits) {
    if (cnt_channel == LEDC_CHANNELS) {
        return; //No channel used for analogWrite
    }
    for (int channel = LEDC_CHANNELS - 1; channel >= cnt_channel; channel--) {
        ledcChangeFrequency(channel, analog_frequency, bits);
    }
    analog_resolution = bits;
}
```
To this:
```
void analogWriteResolution(uint8_t bits) {
    if (cnt_channel != LEDC_CHANNELS) {
        for (int channel = LEDC_CHANNELS - 1; channel >= cnt_channel; channel--) {
            ledcChangeFrequency(channel, analog_frequency, bits);
        }
    }
    analog_resolution = bits;
}
```
So even if no analogWrite was called, next analogWrite will use the analog_resolution set by analogWriteResolution.
And do the same for analogWriteFrequency. This seems to be the best approach.

**Comment by P-R-O-C-H-Y** [comment]
The 2nd part is the error.
The bits resolution can be handled easily, but LEDC works a bit different than standart analogWrite on other ARMs.

Simple rule for using LEDC
lower frequencies -> higher resolution
higher frequencies -> lower resolution (example 40MHz freq can be set only with bit resolution 1)

I was able to set 1Hz frequency using 12 bits resolution ledcSetup(0,1,12);

So we don't know the combinations where it will work or not. Thats where the IDF error appears (there is a calculation for that). I think its like with the touch. Firstly you need to try some touchRead() to see the values of touched/untouched state and than after that you can set some threshold that will be the best for the case. Same here if user wants some frequency and resolution, there is the need of trying that. Most cases will run smooth on the 1st try without changing the resolution. But if you want some maximum or minimum frequencies (1Hz or 40MHz) you will not be able to set them without setting different resolutions.

File: cores/esp32/esp32-hal-ledc.c (hunk 1 above)
[the remark this answers is not in the document]
Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-ledc.c:241]: Check the comments below :)

File: cores/esp32/esp32-hal-ledc.c (hunk 2 above)
[the remark this answers is not in the document]
Comment (P-R-O-C-H-Y) [reply cores/esp32/esp32-hal-ledc.c:252]: Check the comments below :)

**Comment by SuGlider** [comment]
@P-R-O-C-H-Y  - I see your point with `analogWrite()`  and LedC API. Thanks for clarifying. Let's leave the channel management as it is now.

I see these issues right now for the current implementation:

1- `analogWriteResolution()` and `analogWriteFrequency()` have no effect if executed before `analogWrite()`, 
`analog_resolution` and `analog_frequency` are never set in this use case.
2- `analogWriteResolution()` shall verify the resolution parameter and trim it to the maximum allowed by the SoC whenever necessary
3- `analogWriteResolution()` shall verify the frequency parameter and trim it to the maximum allowed by the SoC as well
4- It would be good to issue a warning message when those parameters are invalid, in case they must be adjusted by the Arduino API.

Thanks!

**Comment by vortigont** [comment]
btw, IDF API has some flaws in this area. I.e. duty (read as resolution) can be set but can't be read back by the user code to find out current regs configuration value.
I've hit this earlier working with LEDC espressif/esp-idf#8247

**Comment by SuGlider** [comment]
> btw, IDF API has some flaws in this area. I.e. duty (read as resolution) can be set but can't be read back by the user code to find out current regs configuration value. I've hit this earlier working with LEDC [espressif/esp-idf#8247](https://github.com/espressif/esp-idf/pull/8247)

@vortigont - I understood that you want to know the LEDC maximum resolution for a given channel, right?
Please check the change @P-R-O-C-H-Y has recently done.
There is a parameter `LEDC_MAX_BIT_WIDTH` when including`<driver/ledc.h>` in your sketch or IDF application.
The LEDC maximum resolution is fixed per SoC (ESP32/S2/S3/C3 etc).

**Review by SuGlider — APPROVED** [review]
Great Work!

**Comment by SuGlider** [comment]
@P-R-O-C-H-Y - Do you think that what @vortigont wants is possible to achieve within this PR?

Something like a `getCurrentResolution(ledcChannel)`?

**Comment by SuGlider** [comment]
> @P-R-O-C-H-Y - Do you think that what @vortigont wants is possible to achieve within this PR?
> 
> Something like a `getCurrentResolution(ledcChannel)`?

OK, let's leave it for a future PR. I'll merge it in order to move forward.

**Pull request merged**